src/librustc/ty/layout.rs

```diff
@@ -1762,6 +1771,125 @@ impl<'a, 'tcx, C> TyLayoutMethods<'tcx, C> for Ty<'tcx>
             }
         })
     }
+
+    fn pointee_info_at(this: TyLayout<'tcx>, cx: &C, offset: Size
+    ) -> Option<PointeeInfo> {
+        let mut result = None;
+        match this.ty.sty {
+            ty::RawPtr(mt) if offset.bytes() == 0 => {
+                result = cx.layout_of(mt.ty).ok()
+                    .map(|layout| PointeeInfo {
+                        size: layout.size,
+                        align: layout.align.abi,
+                        safe: None,
+                    });
+            }
+
+            ty::Ref(_, ty, mt) if offset.bytes() == 0 => {
+                let tcx = cx.tcx();
+                let is_freeze = ty.is_freeze(tcx, ty::ParamEnv::reveal_all(), DUMMY_SP);
```

Comment:
I don't know where `pointee_info_at` is called. But depending on where it's used it would make sense to pass in the param env as an argument and then just forward it here instead of assuming `reveal_all`

Reply:
Right now `pointee_info_at` is only used in librustc_codegen_llvm. This pull request would expand that to `librustc_mir`. I'm not familiar with the specifics of [`ParamEnv`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.ParamEnv.html), but judging from the docs it would likely only be used with `ParamEnv::reveal_all` (it is only used after type-checking). So for the current use-case I don't think that extra flexibility is necessary. Would you still prefer to have this parameter in order to make it an explicit caller decision @oli-obk?

Reply:
If we're using it in `librustc_mir`, it is likely that we need to use other `ParamEnv`s. When calling generic const fns during const eval while evaluating an associated constant we'll definitely run into trouble here.

If there are very few use sites, it's not too much hazzle to just pass `ParamEnv::reveal_all()` at them for now

Reply:
Oh right. Curious situation.

You can break this by adding another associated type to https://github.com/rust-lang/rust/blob/79d8a0fcefa5134db2a94739b1d18daa01fc6e9f/src/librustc_target/abi/mod.rs#L899 for the `ParamEnv` and then use that associated type. In rustc you can then insert the correct type

Reply:
An associated type, of course! I have added an associated type `ParamEnv` to `TyLayoutMethods`, the trait that also includes `pointee_info_at`. I think it makes more sense that way as opposed to adding the type to `LayoutOf`, as that last trait does not need the type itself. I can of course move it there if you have good reason to prefer it declared in `LayoutOf` @oli-obk? :wink:

Reply:
Seems fine to me. I just thought it would make sense in `LayoutOf` for consistency. We can still move it there if needed in the future, so the status quo is fine with me